### PR TITLE
DEVX-552: Add Ruby client examples to Confluent Cloud

### PR DESCRIPTION
**Pull request by stanislavkozlovski**
Followed the example from #73 

cc @ybyzek @mhowlett @gAmUssA @edenhill @gwenshap @rnpridgeon for review

### CODE REVIEW

File: clients/cloud/README.md
Comment (ybyzek) [inline clients/cloud/README.md:7]: Resize `images/ruby.png` to be a bit smaller?

**Review by ybyzek — APPROVED** [review]
Thanks for the PR @stanislavkozlovski !

LGTM from my perspective, but please don't merge until review/approval from the clients team.

File: clients/cloud/ruby/README.md
Comment (edenhill) [inline clients/cloud/ruby/README.md:20]: What directly should this be run from?

File: clients/cloud/ruby/c_cloud.rb
Comment (edenhill) [inline clients/cloud/ruby/c_cloud.rb:37]: What is telling the client to connect with SSL?

File: clients/cloud/ruby/c_cloud.rb
Comment (edenhill) [inline clients/cloud/ruby/c_cloud.rb]: Since this runs in the foreground I'd suggest SIGINT (ctrl-c) instead (or both)

File: clients/cloud/ruby/c_cloud.rb
Comment (edenhill) [inline clients/cloud/ruby/c_cloud.rb]: Python -> Ruby

Comment (edenhill) [reply clients/cloud/ruby/c_cloud.rb]: Since this is used for the consumer it probably shouldn't say "produce messages" here

File: clients/cloud/ruby/consumer.rb
Comment (edenhill) [inline clients/cloud/ruby/consumer.rb]: I don't know ruby, so does this exit the loop or continue going? (should be the latter)

File: clients/cloud/ruby/producer.rb
Comment (edenhill) [inline clients/cloud/ruby/producer.rb]: This needs to handle TOPIC_ALREADY_EXISTS, right?

File: clients/cloud/ruby/producer.rb
Comment (edenhill) [inline clients/cloud/ruby/producer.rb]: deliver_message() is a synchronous send, which is fine for this example but generally not what people should be using since it will absolutely kill performance.

I'd like for these examples to "do the right thing" and use an asynchronous producer.

**Review by edenhill — CHANGES_REQUESTED** [review]
Nice stuff, bunch of comments in line.

File: clients/cloud/ruby/README.md
Comment (stanislavkozlovski) [reply clients/cloud/ruby/README.md:20]: From the `/ruby` directory. I'm unsure on how to best specify this

File: clients/cloud/ruby/c_cloud.rb
Comment (stanislavkozlovski) [reply clients/cloud/ruby/c_cloud.rb:37]: the `ssl_ca_certs_from_system` arg

From the zendesk docs:

> Authentication using SASL
Kafka has support for using SASL to authenticate clients. Currently GSSAPI, SCRAM and PLAIN mechanisms are supported by ruby-kafka.

> NOTE: With SASL for authentication, it is highly recommended to use SSL encryption. The default behavior of ruby-kafka enforces you to use SSL and you need to configure SSL encryption by passing ssl_ca_cert or enabling ssl_ca_certs_from_system. However, this strict SSL mode check can be disabled by setting sasl_over_ssl to false while initializing the client.

File: clients/cloud/ruby/c_cloud.rb
Comment (stanislavkozlovski) [reply clients/cloud/ruby/c_cloud.rb]: This actually catches all and is a left-over comment - so I'll just remove it

File: clients/cloud/ruby/c_cloud.rb
Comment (stanislavkozlovski) [reply clients/cloud/ruby/c_cloud.rb]: Oh wow, whoops

File: clients/cloud/ruby/consumer.rb
Comment (stanislavkozlovski) [reply clients/cloud/ruby/consumer.rb]: Ugh, yeah... it seems to re-raise every error. Should we just keep retrying infinitely?

File: clients/cloud/ruby/producer.rb
Comment (stanislavkozlovski) [reply clients/cloud/ruby/producer.rb]: Yeah, we should fail on other errors 👍

File: clients/cloud/ruby/producer.rb
Comment (stanislavkozlovski) [reply clients/cloud/ruby/producer.rb]: I was thinking about this and decided to go with the simpler approach for the example. I don't feel strongly about this - i'll switch to the async producer

File: clients/cloud/ruby/c_cloud.rb
Comment (stanislavkozlovski) [inline clients/cloud/ruby/c_cloud.rb:53]: `||=` initializes `@producer` if it's nil, otherwise doesn't do anything (always returns `@producer`)

File: clients/cloud/ruby/README.md
Comment (edenhill) [inline clients/cloud/ruby/README.md]: Should be librdkafka.config to avoid ambiguity with `~/.ccloud/config` (which is the ccloud tool config file)

File: clients/cloud/ruby/README.md
Comment (edenhill) [reply clients/cloud/ruby/README.md:20]: Add it as a bash step before bundle install, assuming being in repo top dir:
`$ cd clients/ruby`  or whever it is

File: clients/cloud/ruby/README.md
Comment (edenhill) [inline clients/cloud/ruby/README.md]: You don't need to list the output of the command here.

File: clients/cloud/ruby/README.md
Comment (edenhill) [inline clients/cloud/ruby/README.md:43]: I count 11 messages ;)

File: clients/cloud/ruby/c_cloud.rb
Comment (edenhill) [inline clients/cloud/ruby/c_cloud.rb]: can we do better, e.g 1-10ms?

People will cut'n'paste these examples, change them, and eventually be annoyed that ccloud latency is horrible.

**Review by edenhill — CHANGES_REQUESTED** [review]

File: clients/cloud/ruby/README.md
Comment (stanislavkozlovski) [reply clients/cloud/ruby/README.md]: Fair enough. I opted for `ccloud.config` because of this client not being based on librdkafka. This is better for consistency across the examples as well

File: clients/cloud/ruby/README.md
Comment (stanislavkozlovski) [reply clients/cloud/ruby/README.md:20]: This readme is located in `clients/ruby`. Wouldn't users read it from there?

File: clients/cloud/ruby/README.md
Comment (stanislavkozlovski) [reply clients/cloud/ruby/README.md:43]: Good catch, this and the consumer example were outdated - I had fixed it in the code but not here hehe

File: clients/cloud/ruby/c_cloud.rb
Comment (stanislavkozlovski) [reply clients/cloud/ruby/c_cloud.rb]: For sure. It's worth noting that the library gives an example of 30 **seconds**

File: clients/cloud/ruby/README.md
Comment (edenhill) [reply clients/cloud/ruby/README.md]: Oh, right, librdkafka.config is indeed a bad name, but so is ccloud.config.  Maybe example.config ?

File: clients/cloud/ruby/c_cloud.rb
Comment (edenhill) [reply clients/cloud/ruby/c_cloud.rb]: Now that is a crazy default latency

File: clients/cloud/ruby/README.md
Comment (edenhill) [inline clients/cloud/ruby/README.md:12]: @ybyzek Maybe we should use this filename in all of the examples? The files are identical anyway, and the example filename avoids it getting confused with any other config files in ccloud/

Comment (stanislavkozlovski) [reply clients/cloud/ruby/README.md:12]: I also think so. I pushed a commit to do that - cc @ybyzek

File: clients/cloud/ruby/README.md
Comment (edenhill) [reply clients/cloud/ruby/README.md:20]: Maybe, but being explicit here doesn't hurt. People tend to need complete instructions

**Comment by stanislavkozlovski** [comment]
@edenhill  @ybyzek is this good to merge?

**Comment by ybyzek** [comment]
@stanislavkozlovski : LGTM but note there are some conflicts in the `README.md`'s because I modified those in another commit too.

**Comment by stanislavkozlovski** [comment]
Rebased 👍

File: clients/cloud/ruby/c_cloud.rb
Comment (edenhill) [inline clients/cloud/ruby/c_cloud.rb]: I think you can remove the "tweaked .*" part, it is the ruby defaults that are terribly off.

File: clients/cloud/ruby/producer.rb
Comment (edenhill) [inline clients/cloud/ruby/producer.rb:54]: how are delivery errors reported?

Comment (stanislavkozlovski) [reply clients/cloud/ruby/producer.rb:54]: Seems like they're just dropped. The `.shutdown` method actually ensures delivery. Digging into it, it just waits for a Worker thread to shut down when it gives it a `:shutdown` message. Here is how it is handled:
```ruby
when :shutdown
            begin
              # Deliver any pending messages first.
              @producer.deliver_messages
            rescue Error => e
              @logger.error("Failed to deliver messages during shutdown: #{e.message}")

              @instrumenter.instrument("drop_messages.async_producer", {
                message_count: @producer.buffer_size + @queue.size,
              })
            end

            # Stop the run loop.
            break
```
`Error` here is any ruby error

Comment (edenhill) [reply clients/cloud/ruby/producer.rb:54]: So the application would need to check that instrumenter message_count thing to know if messages were dropped?
Isn't there a per-message report/future/whatever?

Comment (stanislavkozlovski) [reply clients/cloud/ruby/producer.rb:54]: Nope
```
If you use the asynchronous producer there's no guarantee that messages will have been delivered after #deliver_messages returns. A way of blocking until a message has been delivered with the asynchronous producer may be implemented in the future.
```

**Comment by stanislavkozlovski** [comment]
Oops, I apparently forgot to push the `tweaked` comment removal commit. Sorry about that.

@edenhill is this good to go?

File: clients/cloud/ruby/producer.rb
Comment (edenhill) [reply clients/cloud/ruby/producer.rb:54]: That's terrible.

**Review by edenhill — APPROVED** [review]
LGTM!

**Comment by ybyzek** [comment]
@stanislavkozlovski , sorry for not noticing this before, but can you please change the destination branch from `master` to `5.0.1-post`?

**Comment by ybyzek** [comment]
I am closing this PR because it was merged to the correct branch with https://github.com/confluentinc/examples/pull/99

**Pull request closed without merging**
